topics/authentication.md

```diff
@@ -6,140 +6,151 @@ Code examples: <a href="samples.md" anchor="authentication">Authentication</a>
 
->The handler won't be executed if the configured authentication fails (when returning `null` in the authentication mechanism)
->
-{type="note"}
+```kotlin
+```
+{src="snippets/auth-basic/src/main/kotlin/com/example/Application.kt" lines="9-20"}
 
-## Naming the authentication provider
+To understand how the `validate` function works, we need to introduce two terms:
+* A _principal_ is an entity that can be authenticated: a user, a computer, a service, etc. In Ktor, various authentication providers might use different principals. For example, the `basic` and `form` providers authenticate [UserIdPrincipal](https://api.ktor.io/ktor-features/ktor-auth/ktor-auth/io.ktor.auth/-user-id-principal/index.html) while the `jwt` provider verifies [JWTPrincipal](https://api.ktor.io/ktor-features/ktor-auth-jwt/ktor-auth-jwt/io.ktor.auth.jwt/-j-w-t-principal/index.html).
+  > If you use [session authentication](session-auth.md), a principal might be a data class that stores session data.
+* A _credential_ is a set of properties for a server to authenticate a principal: a user/password pair, an API key, and so on. For instance, the `basic` and `form` providers use [UserPasswordCredential](https://api.ktor.io/ktor-features/ktor-auth/ktor-auth/io.ktor.auth/-user-password-credential/index.html) to validate a user name and password while `jwt` validates [JWTCredential](https://api.ktor.io/ktor-features/ktor-auth-jwt/ktor-auth-jwt/io.ktor.auth.jwt/-j-w-t-credential/index.html).
 
-It is possible to give arbitrary names to the authentication providers you specify,
-or to not provide a name at all (unnamed provider) by not setting the name argument or passing a null.
+So, the `validate` function checks a specified [Credential](https://api.ktor.io/ktor-features/ktor-auth/ktor-auth/io.ktor.auth/-credential/index.html) and returns a [Principal](https://api.ktor.io/ktor-features/ktor-auth/ktor-auth/io.ktor.auth/-principal/index.html) in a case of successful authentication or `null` if authentication fails.
 
-You cannot repeat authentication provider names, and you can define just one provider without a name.
+> To skip authentication based on a specific criteria, use [skipWhen](https://api.ktor.io/ktor-features/ktor-auth/ktor-auth/io.ktor.auth/-authentication-provider/skip-when.html). For example, you can skip `basic` authentication if a [session](sessions.md) already exists:
+> ```kotlin
+> basic {
+>     skipWhen { call -> call.sessions.get<UserSession>() != null }
+> }
 
-In the case you repeat a name for the provider or try to define two unnamed providers, an exception will be thrown:
 
-```text
-java.lang.IllegalArgumentException: Provider with the name `authName` is already registered
-```
+### Step 4: Protect specific routes {id="authenticate-route"}
 
-Summarizing:
+After defining and configuring an authentication provider, you can protect specific [routes](Routing_in_Ktor.md) using the
```

Comment:
After defining and configuring an authentication provider, you can protect specific routes using the -> 

The final step is to define the authorization for the different resources in our application.